### ISSUE TITLE
The results of SafeDrug model differ significantly from those in the paper.

### ISSUE THREAD

**Issue by He-Yichen** [sunlabuiuc/PyHealth#220]
Hi! sorry for bothering you again.
I ran the code for GAMENet, SafeDrug and MoleRec locally. The results of the three models are as follows:
![394bb702054f7c354d1e4074b664a4a](https://github.com/sunlabuiuc/PyHealth/assets/90712879/91b23528-7b9a-4412-8297-ae43694e0669)
Here is the problem: the jaccard_samples of my local SafeDrug can only reach about 0.33. Theoretically, the jaccard_samples of SafeDrug should be similar to GAMENet. Why is there such a big gap? 
Additionally,  why the results obtained with pyhealth are lower than in the paper? I note that the sample dataset contains 14,142 visits and 5,449 patients, which is different from the papers that contain 6,350 patients and 14,995 visits. Is it because of this?
![image](https://github.com/sunlabuiuc/PyHealth/assets/90712879/3273772c-5287-4d59-ae28-1ee8625feda0)
Looking forward to and thank you for your reply!

**Comment by ycq091044** [comment]
Hello, could you send the code you used for training and inference for these three models?

**Comment by He-Yichen** [comment]
> Hello, could you send the code you used for training and inference for these three models?

Sure, my code is the same as **examples/drug_recommendation_mimic3_gamenet.py**, **drug_recommendation_mimic3_safedryg.py** and **drug_recommendation_mimic3_molerec.py**.

Here's my data processing code, which is is the same as example.
![image](https://github.com/sunlabuiuc/PyHealth/assets/90712879/795cdc7c-1712-4cff-af4a-82eb6f3e7c28)

The implementation of three models is:
```python
from pyhealth.models import MoleRec, SafeDrug, GAMENet
``` 
The code I used for training and inference for these three models are as flow:
```python
#GAMENet
gameNetModel = GAMENet(
    sample_dataset,
)
gameNetTrainer = Trainer(model=gameNetModel)
gameNetTrainer.train(
    train_dataloader=train_dataloader,
    val_dataloader=val_dataloader,
    epochs=100,
    monitor="pr_auc_samples",
)
#SafeDrug
safeDrugModel = SafeDrug(
    sample_dataset,
)
safeDrugTrainer = Trainer(model=safeDrugModel)
safeDrugTrainer.train(
    train_dataloader=train_dataloader,
    val_dataloader=val_dataloader,
    epochs=100,
    monitor="pr_auc_samples",
)
#MoleRec
moleRecModel = MoleRec(
    sample_dataset,
    feature_keys=["conditions", "procedures"],
    label_key="drugs",
    mode="multilabel",
)
trainer = Trainer(model=moleRecModel ,device="cuda")
trainer.train(
    train_dataloader=train_dataloader,
    val_dataloader=val_dataloader,
    epochs=100,
    monitor="pr_auc_samples",
)
``` 
Additionally, Mr. [Tyunsen](https://github.com/Tyunsen) seems to have the same confusion as me:
https://github.com/sunlabuiuc/PyHealth/issues/209#issuecomment-1722465032
Looking forward to your reply! @ycq091044

**Comment by ycq091044** [comment]
Hello @He-Yichen , I find the SafeDrug adaptive loss has issues and already solved it. Please install the new pyhealth version by "git clone" and "pip install ." Thanks!

**Comment by He-Yichen** [comment]
Thanks for deal with my problem !
